### DISCUSSION TITLE
Multiple providers on same host

### DISCUSSION THREAD

**Discussion by brandonk10** [OpenIDC/mod_auth_openidc#737] [Q&A]
I have 2 realms working fine on the same virtual host, at least independently.  If I log into one, any directory protected by the other fails as unauthorized.  This is because they, by default, share the same session cookie.  OIDCCookie is configurable at the directory level.  However, unless I missed a step, all providers need to share the same redirect_uri, which is only at virtual host level.  Its the redirect_uri that actually sets the session cookie.  I'm probably missing something here, but, if not, would it make sense to allow redirect_uri to be configured/overridden at directory level or in the metadata configuration?

**Comment by zandbelt** [comment]
yes, the redirect URI would be the same for both providers; why would you need to be logged in at 2 providers at the same time?

**Reply by brandonk10** [reply]
If I have 2 userbases for each application, administered independently, there could be overlap between the two, so this might be possible.  If I had single administrator of the users, I could use authorization to split between the applications, but I haven't seen a way to split administration in the same realm.

**Reply by zandbelt** [reply]
sure, but not in a single browser; for your setup you can just use a single Redirect URI

**Reply by brandonk10** [reply]
users just get URLs, they won't know to use new session tabs, etc.  their experience would be in the same browser.  Admittedly, this is not a common use case on the same host.

**Reply by zandbelt** [reply]
then you'd have to use different vhosts

**Reply by brandonk10** [reply]
I understand this is how the module currently operates.  Is there any philosophical objection to allowing redirect_uri at the provider level?

**Reply by studersi** [reply]
Just as an FYI: We have had similar use cases in the past with CAS authentication: https://github.com/apereo/mod_auth_cas/pull/160. So it might not be that uncommon.

**Reply by brandonk10** [reply]
FWIW, I'm working on a fork for changes to allow some of these items at directory level.  I'm not sure if provider-level config would work - it seems like redirect_uri is needed in the flow before the provider is determined.  Right now, the only thing I haven't figured out is where to perform the configuration validation as there is no post_config hook for directory-level configuration.  I'm hooking content handler now, but I think that is too late in the flow.
